File: elixir/2015/day07a/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  import Bitwise

  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp is_integer?(str) do
    case Integer.parse(str) do
      {_, ""} -> true
      _ -> false
    end
  end

  defp eval(operations, cache, expr) do
    cond do
      is_integer?(expr) ->
        {int_val, _} = Integer.parse(expr)
        {int_val, cache}
      Map.has_key?(cache, expr) ->
        {Map.fetch!(cache, expr), cache}
      true ->
        {val, new_cache} = 
          case Map.fetch!(operations, expr) do
            {:assign, src} ->
              {val1, cache1} = eval(operations, cache, src)
              {val1 &&& 0xffff, cache1}
            {:not, src} ->
              {val1, cache1} = eval(operations, cache, src)
              {~~~val1 &&& 0xffff, cache1}
            {:and, src1, src2} ->
              {val1, cache1} = eval(operations, cache, src1)
              {val2, cache2} = eval(operations, cache1, src2)
              {val1 &&& val2 &&& 0xffff, cache2}
            {:or, src1, src2} ->
              {val1, cache1} = eval(operations, cache, src1)
              {val2, cache2} = eval(operations, cache1, src2)
              {val1 ||| val2 &&& 0xffff, cache2}
            {:lshift, src, amt} ->
              {val1, cache1} = eval(operations, cache, src)
              {(val1 <<< amt) &&& 0xffff, cache1}
            {:rshift, src, amt} ->
              {val1, cache1} = eval(operations, cache, src)
              {(val1 >>> amt) &&& 0xffff, cache1}
          end
        updated_cache = Map.put(new_cache, expr, val)
        {val, updated_cache}
    end
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        regexes = [
          {:assign, ~r/^(\d+|\w+) -> (\w+)$/},
          {:not, ~r/^NOT (\d+|\w+) -> (\w+)$/},
          {:and, ~r/(\d+|\w+) AND (\d+|\w+) -> (\w+)$/},
          {:or, ~r/(\d+|\w+) OR (\d+|\w+) -> (\w+)$/},
          {:lshift, ~r/(\d+|\w+) LSHIFT (\d+) -> (\w+)$/},
          {:rshift, ~r/(\d+|\w+) RSHIFT (\d+) -> (\w+)$/},
        ]
        operations =
          content
          |> String.split(~r/\r?\n/)
          |> Enum.reduce(%{}, fn line, curr_ops ->
            Enum.reduce_while(regexes, curr_ops, fn {op_type, re}, acc ->
              case Regex.run(re, line) do
                nil ->
                  {:cont, acc}
                [_, src, dest] when op_type == :assign or op_type == :not ->
                  {:halt, Map.put(acc, dest, {op_type, src})}
                [_, src1, src2, dest] when op_type in [:and, :or] ->
                  {:halt, Map.put(acc, dest, {op_type, src1, src2})}
                [_, src, amtStr, dest] when op_type in [:lshift, :rshift] ->
                  {amt, _} = Integer.parse(amtStr)
                  {:halt, Map.put(acc, dest, {op_type, src, amt})}
                _ ->
                  {:cont, acc}
              end
            end)
          end)
        {total, _} = eval(operations, %{}, "a")
        {:ok, total}
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: fsharp/2015/day07a/Program.fs
open System
open System.IO
open System.Text.RegularExpressions

type Operation =
   | Assign of string
   | Not of string
   | And of string * string
   | Or of string * string
   | LeftShift of string * int
   | RightShift of string * int

let rec eval (operations: Map<string, Operation>) (cache: Map<string, int>) (expr: string) : int * Map<string, int> =
   match Int32.TryParse(expr) with
   | true, n -> n, cache
   | false, _ ->
       match Map.tryFind expr cache with
       | Some value -> value, cache
       | None ->
           let op = Map.find expr operations
           let eval' e c = eval operations c e
           let result, newCache =
               match op with
               | Assign src ->
                   let res, c = eval' src cache
                   res, c
               | Not src ->
                   let res, c = eval' src cache
                   ~~~res, c
               | And (src1, src2) ->
                   let r1, c1 = eval' src1 cache
                   let r2, c2 = eval' src2 c1
                   r1 &&& r2, c2
               | Or(src1, src2) ->
                   let r1, c1 = eval' src1 cache
                   let r2, c2 = eval' src2 c1
                   r1 ||| r2, c2
               | LeftShift(src, amt) ->
                   let res, c = eval' src cache
                   res <<< amt, c
               | RightShift(src, amt) ->
                   let res,c  = eval' src cache
                   res >>> amt, c
           let masked = result &&& 0xffff
           masked, Map.add expr masked newCache

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processFile(filename: string) : int =
    let regexes : (Regex * (Match -> string * Operation)) list =
        [
            Regex(@"^(\d+|\w+) -> (\w+)$"),
            fun m -> m.Groups.[2].Value, Assign m.Groups.[1].Value

            Regex(@"NOT (\d+|\w+) -> (\w+)$"),
            fun m -> m.Groups.[2].Value, Not m.Groups.[1].Value

            Regex(@"(\d+|\w+) AND (\d+|\w+) -> (\w+)$"),
            fun m -> m.Groups.[3].Value, And (m.Groups.[1].Value, m.Groups.[2].Value)

            Regex(@"(\d+|\w+) OR (\d+|\w+) -> (\w+)$"),
            fun m -> m.Groups.[3].Value, Or (m.Groups.[1].Value, m.Groups.[2].Value)

            Regex(@"(\d+|\w+) LSHIFT (\d+) -> (\w+)$"),
            fun m -> m.Groups.[3].Value, LeftShift (m.Groups.[1].Value, int m.Groups.[2].Value)

            Regex(@"(\d+|\w+) RSHIFT (\d+) -> (\w+)$"),
            fun m -> m.Groups.[3].Value, RightShift (m.Groups.[1].Value, int m.Groups.[2].Value)
        ]
    let operations : Map<string, Operation> = 
        File.ReadAllLines(filename)
        |> Array.fold (fun acc line ->
            let matched =
                regexes
                |> List.tryPick (fun (rx, handler) ->
                    let m = rx.Match(line)
                    if m.Success then Some (handler m) else None
                )
            match matched with
            | Some (key, op) -> Map.add key op acc
            | None ->
                eprintfn "error: malformed line: %s" line
                Environment.Exit(1)
                acc
        ) Map.empty
    let result, _ = eval operations Map.empty "a"
    result

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
